### ISSUE TITLE
write/read nodes / load save levels

### ISSUE THREAD

**Issue by zartan9** [armory3d/armory#2070]
hi im amazed at all the features of armory! figuring out level loading (through all your support), now saving... I looked around a bit, but cant find a working way to load game levels (scenes) if there is a link to it please let me know , also would it work for all export options (mobile/desktop etc). Thanks happy holidays

**Comment by MoritzBrueckner** [comment]
Hi,

there are a bunch of options (and they are *badly/not* documented 😨):

- [Read/Write File node](https://github.com/armory3d/armory/wiki/reference#write-file):
  Reads or writes a string from a file, looking in the code revealed that it's only implemented in Krom. I guess it should be possible to implement this for other targets as well, html5 might be a bit tricky because of browser restrictions. With this node you would need to impement all the decoding/encoding of the level into a string yourself.

- [Read/Write JSON node](https://github.com/armory3d/armory/wiki/reference#write-json):
  Lets you read/write .json files. This is a bit easier to use as you can use it to store types like (nested) dynamic arrays. However, this is Krom only as well :(

- [Read/Write Storage node](https://github.com/armory3d/armory/wiki/reference#write-storage):
  Those nodes should work on all platforms as they rely on a Kha API. They allow you to access a dictionary/map (key value pairs) that (I guess?) is stored between sessions. It might be a bit tricky/tedious though to store levels in there, but for just storing some informations about the player character for example should be very fast to implement for you.

There is no generalized built-in way of storing levels because it depends from game to game what should be saved and what not.

I will add a `docs` label so that we know that the documentation must be improved for this. Also we should think about implementing those nodes for other targets and make them more intuitive to use.

Thanks :)

**Comment by zartan9** [comment]
thanks! Wow Armory is like a diamond in the rough...so much power but trying to harness it takes time!! I think ill go with option 3 since id like to make game for many formats... i found the right nodes... not sure of "dictionary/map/key" thing... Im getting confused on Keyboard (press "b" to load) with "key" is this the "scene001" or some string to type in to load... i like visual scripting nodes since i dont really know C# etc but i guess ill learn some things
![load](https://user-images.githubusercontent.com/74879950/103185659-5645c880-488b-11eb-8685-21c224981013.png)
 as i go along.. Screenshot attached..

**Comment by MoritzBrueckner** [comment]
Imagine a bunch of boxes with names. Those names are called _keys_. In each box there can be one and only one item, which is called a _value_. If you set a value and there already exists one, the old value is overridden. So for example you would save things like: `player_health` as a key with `100` being the value. Or `player_position` being `[5.0, 10.0, 1.0]` for example.

A common name for that box/item thing is a _map_ (this is how it is called in Haxe or Java for example) or a _dictionary_ (Python e.g.).

However, such questions are better suited for the forum, Github is mainly used for bugs and feature requests. But please let this issue open for know until there is at least some documentation about the current state of those nodes mentioned above :)

**Comment by zartan9** [comment]
ok ill try to wrap my head around this! ...for a scene loader, the map would be  "current_scene" with the key of 1... (First level/scene)...   and next level would have key of 2 (2nd level)... and each value is auto over ridden so whereever the player
stops playing, the latest level is the last value (2 for example) so... when it loads (pressing button b) current_scene i dont have to put anything in the default (its automatic/ like leaving cookies or something??) sorry if what im saying doesnt make sense just trying to figure it. thanks!

**Comment by MoritzBrueckner** [comment]
> for a scene loader, the map would be "current_scene" with the key of 1

Not really, the map is the entire set of key/value pairs (I improved my answer above). The key would be `current_scene` and the value could be `1` or `2` depending on the scene for example.

> and each value is auto over ridden so whereever the player
> stops playing

That depends on what you program. If you update the value each time the player moves, then yes (like auto save, but better do something like that in intervals for better overall performance). You could also only set that value if the player clicks on a `Save` button for example.

If you then load the game you have to read that value and set the current scene according to that (it's a lot of manual work...).

> like leaving cookies or something??

Yes pretty much like that. If I'm not mistaken the storage system uses local storage on the html5 target which is similar.

**Comment by QuantumCoderQC** [comment]
@zartan9 Hi, If you like to load a level/scene, then use the `Spawn Scene` [node](https://github.com/armory3d/armory/wiki/reference#spawn-scene) . This is when you have already created multiple game levels with different objects placements, materials, and so on in different Blender scenes. Then during the gameplay, you could use `Remove Scene Active` ([node](https://github.com/armory3d/armory/wiki/reference#remove-scene-active)) to remove the current scene and then spawn the new scene. 

If you like to store player information(Ex: Health, Name, etc.), statistics(Ex: High Scores, time, etc), transforms of certain objects and so on, you could use the `Read/Write storage Node`

Hope this helps.

**Comment by zartan9** [comment]
Thanks everyone for clarifying but sounds like this may be over my head (Im more of a modeller)... can there be a simple working example uploaded to see how when pressing "load" button it loads the "spawned" scene/last played level?...

**Comment by MoritzBrueckner** [comment]
[LoadSaveScene.zip](https://github.com/armory3d/armory/files/5751489/LoadSaveScene.zip)

Made you a small example file, you can switch scenes with "1" and "2" and there is auto-save. However, it doesn't seem to work on Krom (in html5 it works). Also setting the scene on init didn't work (needs to be fixed) so I load it after a small delay.

**Comment by QuantumCoderQC** [comment]
@MoritzBrueckner  Thanks for the file. The interesting thig is when the example is run in Krom, there is in fact a file called `default.kha` created under `C:\Users\USER\Saved Games\LoadSaveScene` on Windows. 

When opened with notepad, it reads `oy8:scene_idi2g`. This is the case when game was closed with Scene 2 active.

However, it does not seem to load the file when the game starts again.

**Comment by MoritzBrueckner** [comment]
Interesting. I printed out the values and it even correctly reads the storage, there must be some other issue.

**Comment by zartan9** [comment]
hey thanks! I pressed scene 2, and when compiled again, it loaded as scene 2. but after that i loaded it it went back to1! I also then did the same thing (pressed to scene 2) and it was 2  twice, but then went back to 1!!! Im on firefox! at least i can see how it is done (integers!?!) thanks.

**Comment by MoritzBrueckner** [comment]
I guess its a case of a [race condition](https://stackoverflow.com/questions/34510/what-is-a-race-condition), the outcome depends on whether Iron first initializes all traits or first sets the default scene. Increasing the timer could help, but its not a good solution, only an ugly workaround.

> at least i can see how it is done (integers!?!)

You can use integers, yes, but all primitive data types such as floats, booleans, strings (not really a primitive type though) should work. Arrays and maps should work as well.

**Comment by zartan9** [comment]
wow! complex stuff!! i heard of arrays i think that may work best if i have many levels... ill look into armory documentation on arrays next thanks!! happy new year soon

**Comment by knowledgenude** [comment]
Case someone would like to know it in the docs, In Linux the directory used by Read/Write storage is different. It is in a hidden folder named as the .blend file inside `Home` folder.

@MoritzBrueckner,  i guess this issue with on init can be also solved with your refactor, so i am leaving this reminder.
